### PR TITLE
HA tests

### PR DESCRIPTION
**Pull request by dragomirp**

### CODE REVIEW

File: charmcraft.yaml [hunk 1]
```diff
@@ -17,11 +17,3 @@ bases:
     run-on:
     - name: "ubuntu"
       channel: "20.04"
-parts:
```
Comment (dragomirp) [inline charmcraft.yaml:20]: No longer needed

File: tests/integration/ha_tests/application_charm/src/continuous_writes.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# Copyright 2022 Canonical Ltd.
```
Comment (dragomirp) [inline tests/integration/ha_tests/application_charm/src/continuous_writes.py:1]: Taken from the mongodb-operator charm

File: tests/integration/ha_tests/application_charm/src/charm.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
+#!/usr/bin/env python3
```
Comment (dragomirp) [inline tests/integration/ha_tests/application_charm/src/charm.py:1]: Mostly taken from mysql-k8s-operator charm

File: tox.ini [hunk 4]
```diff
@@ -61,10 +61,15 @@ commands =
...
+[testenv:ha-integration]
+description = Run HA integration tests
+commands =
+    poetry install
+    poetry export -f requirements.txt -o requirements.txt
+    poetry run pytest -v --tb native {[vars]tst_path}integration/ha_tests --log-cli-level=INFO -s {posargs} --durations=0
 
 [testenv:tls-integration]
 description = Run tls integration tests
-deps =
```
Comment (dragomirp) [inline tox.ini:67]: So we don't need to install poetry separately

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -67,8 +91,6 @@ jobs:
           bootstrap-options: "--agent-version 2.9.29"
       - name: Install tox
         run: python3 -m pip install tox
-      - name: Install poetry
```
Comment (dragomirp) [inline .github/workflows/ci.yaml:70]: Get poetry as tox dep

File: tests/integration/ha_tests/helpers.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+        container_name: The name of the container to send signal to
+        process: The name of the process to send signal to
+        signal_code: The code of the signal to send
+    """
+    kubernetes.config.load_kube_config()
+
+    pod_name = unit_name.replace("/", "-")
+
+    send_signal_command = f"pkill --signal {signal_code} -f {process}"
+    response = kubernetes.stream.stream(
```
Comment (WRFitch) [inline tests/integration/ha_tests/helpers.py:249]: This is neat, but consider using `ops_test.juju`:
```python
cmd = f"ssh --container {container_name} {unit_name} pkill --signal {signal_code} -f {process}"
ops_test.juju(*cmd.split(" "))
```

File: tests/integration/ha_tests/helpers.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+    Args:
+        ops_test: The ops test framework
+        unit: The unit on which to execute the action on
+
+    Returns:
+        A dictionary representing the cluster status
+    """
+    get_cluster_status_action = await unit.run_action("get-cluster-status")
+    cluster_status_results = await get_cluster_status_action.wait()
+    return cluster_status_results.results
```
Comment (delgod) [inline tests/integration/ha_tests/helpers.py:280]: I don't think this charm has this action. I don't think we should have it. please remove

File: tests/integration/ha_tests/test_ha.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+    clear_writes_action = await application_unit.run_action("clear-continuous-writes")
+    await clear_writes_action.wait()
+
+    start_writes_action = await application_unit.run_action("start-continuous-writes")
+    await start_writes_action.wait()
+
+    yield
+
+    clear_writes_action = await application_unit.run_action("clear-continuous-writes")
```
Comment (delgod) [inline tests/integration/ha_tests/test_ha.py:49]: Should we stop writing before clear? For example, if the test failed, maybe we should stop writes and cleanup data after the test.

File: tests/integration/ha_tests/application_charm/src/charm.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+        """Returns the count of rows in the continuous writes table."""
+        if not self._database_config:
+            return -1
+
+        client = MongoClient(self._database_config["uris"])
+        db = client[DATABASE_NAME]
+        test_collection = db["test_collection"]
+        last_written_value = test_collection.find_one(sort=[("number", -1)])
+        client.close()
+        return last_written_value["number"]
```
Comment (delgod) [inline tests/integration/ha_tests/application_charm/src/charm.py:152]: The last written value should be taken from `src/continuous_writes.py` (not from the database).
For example, `src/continuous_writes.py` can think that it is written 100500 records successfully, but only 500 were written.
Keep in mind the tests below rely on this output.

File: tests/integration/ha_tests/helpers.py [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+    ]
+    return_code, pid, _ = await ops_test.juju(*get_pid_commands)
+
+    assert (
+        return_code == 0
+    ), f"Failed getting pid, unit={unit_name}, container={container_name}, process={process}"
+
+    stripped_pid = pid.strip()
+    if not stripped_pid:
+        return -1
```
Comment (MiaAltieri) [inline tests/integration/ha_tests/helpers.py:227]: I think it would be better to raise an error here, AFAICT, this edge case of `-1` being returned isn't handled

File: tests/integration/ha_tests/helpers.py [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+    Returns:
+        A dictionary representing the cluster status
+    """
+    get_cluster_status_action = await unit.run_action("get-cluster-status")
+    cluster_status_results = await get_cluster_status_action.wait()
+    return cluster_status_results.results
+
+
+def host_to_unit(host: str) -> str:
+    return "/".join(host.split(".")[0].rsplit("-", 1))
```
Comment (MiaAltieri) [inline tests/integration/ha_tests/helpers.py:284]: nit: add docstring

File: tests/integration/ha_tests/helpers.py [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+        return False
+    finally:
+        client.close()
+
+    return True
+
+
+async def get_replica_set_primary(ops_test: OpsTest) -> str:
+    rs_status = await run_mongo_op(ops_test, "rs.status()")
+    assert rs_status.succeeded, "mongod had no response for 'rs.status()'"
```
Comment (MiaAltieri) [inline tests/integration/ha_tests/helpers.py:305]: neat

File: tests/integration/ha_tests/test_ha.py [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+    await send_signal_to_pod_container_process(
+        ops_test,
+        primary,
+        MONGODB_CONTAINER_NAME,
+        MONGOD_PROCESS_NAME,
+        "SIGCONT",
+    )
+
+    # check this after un-freezing the old primary so that if this check fails we still "turned
+    # back on" the mongod process
```
Comment (MiaAltieri) [inline tests/integration/ha_tests/test_ha.py:179]: nice

**Review by MiaAltieri — COMMENTED** [review]
Looks awesome, minor comments only

File: tests/integration/ha_tests/test_ha.py (hunk 8 above)
Comment (dragomirp) [reply tests/integration/ha_tests/test_ha.py:49]: `clear-continuous-writes` action should also call `_on_stop_continuous_writes_action()`, so it should stop it on its own.

File: tests/integration/ha_tests/helpers.py [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APPLICATION_DEFAULT_APP_NAME = "application"
+TIMEOUT = 15 * 60
+
+mongodb_charm, application_charm = None, None
+
+
+async def get_application_name(ops_test: OpsTest, application_name: str) -> str:
+    """Returns the name of the application witt the provided application name.
```
[the remark this answers is not in the document]
Comment (dragomirp) [reply tests/integration/ha_tests/helpers.py]: We use this to get the test app as well, the specific application is passed on as a parameter

File: tests/integration/ha_tests/helpers.py [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+async def deploy_and_scale_application(ops_test: OpsTest) -> str:
+    """Deploys and scales the test application charm.
+
+    Args:
+        ops_test: The ops test framework
+    """
+    application_name = await get_application_name(ops_test, "application")
+
+    if application_name:
+        if len(ops_test.model.applications[application_name].units) != 1:
```
[the remark this answers is not in the document]
Comment (dragomirp) [reply tests/integration/ha_tests/helpers.py]: Moved the common checks and assertions to `scale_application()`

File: tests/integration/ha_tests/helpers.py (hunk 7 above)
Comment (dragomirp) [reply tests/integration/ha_tests/helpers.py]: Removed copypasta

File: tests/integration/ha_tests/application_charm/src/charm.py [hunk 16]
```diff
@@ -120,36 +123,24 @@ def _stop_continuous_writes(self) -> Optional[int]:
         if not self.app_peer_data.get(PROC_PID_KEY):
             return None
 
-        # Send a SIGKILL to the process and wait for the process to exit
-        proc = subprocess.Popen(["pkill", "--signal", "SIGKILL", "-f", "src/continuous_writes.py"])
-        proc.communicate()
+        # Send a SIGTERM to the process and wait for the process to exit
+        os.kill(int(self.app_peer_data[PROC_PID_KEY]), signal.SIGTERM)
```
Comment (dragomirp) [inline tests/integration/ha_tests/application_charm/src/charm.py:127]: We don't really need to call an external process.

File: tests/integration/ha_tests/application_charm/src/continuous_writes.py [hunk 17]
```diff
@@ -2,17 +2,25 @@
...
+def sigterm_handler(_signo, _stack_frame):
+    global run
+    run = False
+
 
 def continous_writes(connection_string: str, starting_number: int):
     write_value = starting_number
 
-    while True:
+    while run:
```
Comment (dragomirp) [inline tests/integration/ha_tests/application_charm/src/continuous_writes.py:23]: Loop until we get a SIGTERM

File: tests/integration/ha_tests/application_charm/src/continuous_writes.py [hunk 18]
```diff
@@ -44,6 +52,9 @@ def continous_writes(connection_string: str, starting_number: int):
 
         write_value += 1
 
+    with open("/tmp/last_written_value", "w") as fd:
```
Comment (dragomirp) [inline tests/integration/ha_tests/application_charm/src/continuous_writes.py:55]: Write the last actually written value to a file for the charm to read

File: tests/integration/ha_tests/application_charm/src/charm.py [hunk 19]
```diff
@@ -188,14 +179,13 @@ def _on_start_continuous_writes_action(self, _) -> None:
...
-            return event.set_results({"writes": 0})
+            return event.set_results({"writes": -1})
 
         writes = self._stop_continuous_writes()
-        event.set_results({"writes": writes})
+        event.set_results({"writes": writes or -1})
 
     def _on_database_created(self, _) -> None:
         """Handle the database created event."""
-        self._start_continuous_writes(1)
```
Comment (dragomirp) [inline tests/integration/ha_tests/application_charm/src/charm.py]: Would be stopped by the fixture either way

File: tests/integration/ha_tests/application_charm/src/charm.py [hunk 20]
```diff
@@ -120,36 +123,24 @@ def _stop_continuous_writes(self) -> Optional[int]:
...
         del self.app_peer_data[PROC_PID_KEY]
 
-        # Query and return the max value inserted in the database
-        # (else -1 if unable to query)
+        # read the last written_value
         try:
             for attempt in Retrying(stop=stop_after_delay(60), wait=wait_fixed(5)):
                 with attempt:
-                    last_written_value = self._max_written_value()
+                    with open(LAST_WRITTEN_FILE, "r") as fd:
```
Comment (dragomirp) [inline tests/integration/ha_tests/application_charm/src/charm.py:135]: It may take a bit for the `continuous_writes` loop to exit, so I recycled the retrying.

File: tests/integration/ha_tests/test_ha.py [hunk 21]
```diff
@@ -125,6 +125,7 @@ async def test_kill_db_process(ops_test: OpsTest, continuous_writes):
     stop_writes_action = await application_unit.run_action("stop-continuous-writes")
     await stop_writes_action.wait()
     total_expected_writes = int(stop_writes_action.results["writes"])
+    assert total_expected_writes > 0, "error while getting total writes."
```
Comment (dragomirp) [inline tests/integration/ha_tests/test_ha.py:128]: Make sure that the test charm actually wrote something and didn't simply error out.

File: tests/integration/ha_tests/test_ha.py [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest) -> None:
+    """Build and deploy three units of MongoDB and one test unit."""
+    # it is possible for users to provide their own cluster for HA testing. Hence check if there
+    # is a pre-existing cluster.
+    mongodb_application_name = await get_application_name(ops_test, APP_NAME)
+    if not mongodb_application_name:
+        mongodb_application_name = await deploy_and_scale_mongodb(ops_test)
```
Comment (delgod) [inline tests/integration/ha_tests/test_ha.py:60]: just Q: is it possible to run this test against currently running MongoDB k8s charm?

File: tests/integration/ha_tests/helpers.py [hunk 23]
```diff
@@ -0,0 +1,361 @@
...
+    for application in ops_test.model.applications:
+        # note that format of the charm field is not exactly "mongodb" but instead takes the form
+        # of `local:focal/mongodb-6`
+        if application_name in status["applications"][application]["charm"]:
+            return application
+
+    return None
+
+
+async def scale_application(
```
[the remark this answers is not in the document]
Comment (delgod) [reply tests/integration/ha_tests/helpers.py:40]: up to @MiaAltieri 
We should have similarly structured tests in MongoDB VM and MongoDB K8s.

**Review by delgod — APPROVED** [review]

File: tests/integration/ha_tests/helpers.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply tests/integration/ha_tests/helpers.py]: Ah, I see. It's still a little confusing. How about:
"""Returns the Application in the juju model that matches the provided application name.

File: tests/integration/ha_tests/test_ha.py (hunk 22 above)
Comment (dragomirp) [reply tests/integration/ha_tests/test_ha.py:60]: Yes, that's how I do it locally to avoid rebuilding the charm on each run.

File: tests/integration/ha_tests/helpers.py [hunk 24]
```diff
@@ -0,0 +1,361 @@
...
+mongodb_charm, application_charm = None, None
+
+
+async def get_application_name(ops_test: OpsTest, application_name: str) -> str:
+    """Returns the name of the application witt the provided application name.
+
+    This enables us to retrieve the name of the deployed application in an existing model.
+
+    Note: if multiple applications with the application name exist,
+    the first one found will be returned.
```
[the remark this answers is not in the document]
Comment (dragomirp) [reply tests/integration/ha_tests/helpers.py:27]: Linting complains if I put the lines together

[the remark this answers is not in the document]
Comment (MiaAltieri) [reply tests/integration/ha_tests/helpers.py:27]: Makes sense, the wrapping is still agressive. Can you make the first line the first 99 chars and the second one the remaining chars?

File: tests/integration/ha_tests/helpers.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply tests/integration/ha_tests/helpers.py:40]: To keep things structured similarly between charms lets keep it as is.

**Review by MiaAltieri — DISMISSED** [review]
Looks good to me, minor things only

**Review by WRFitch — APPROVED** [review]
